buildtest/modules/spider.py

```diff
@@ -0,0 +1,103 @@
+import json
+import os
+import subprocess
+
+from buildtest.tools.defaults import BUILDTEST_SPIDER_FILE, BUILDTEST_CONFIG_FILE
+from buildtest.tools.config import load_configuration
+
+
+class Spider:
+    """Class declaration of Spider class"""
+
+    def __init__(self, tree=None):
+        """Initialize method for Spider class.
+
+        :param tree: User can specify a module tree to query from spider.
+        :type tree: str
+        """
+
+        #  if user specifies a tree, then run spider command otherwise read from configuration file.
+        if tree:
+            self.tree = tree
+            spider_cmd = f"{os.getenv('LMOD_DIR')}/spider -o spider-json {self.tree}"
+
+            out = subprocess.check_output(spider_cmd, shell=True).decode("utf-8")
+            self.spider_content = json.loads(out)
+
+        else:
+            with open(BUILDTEST_SPIDER_FILE, "r") as fd:
+                self.spider_content = json.load(fd)
+
+            content = load_configuration(BUILDTEST_CONFIG_FILE)
+            self.tree = content["BUILDTEST_MODULEPATH"]
+
+    def get_trees(self):
+        """" Return module trees used in spider command
+
+        :return: return module trees used for querying from spider
+        :rtype: str
+        """
+        return self.tree
+
+    def get_unique_software(self):
+        """Return all keys from spider. This is the unique software names.
+
+        :return: return sorted list of all spider keys.
+        :rtype: list
+        """
+
+        return sorted(list(self.spider_content.keys()))
+
+    def get_modules(self):
+        """Retrieve all module names from all module tree.
+
+        :return: returns  a sorted list of all full canonical module name from all spider records.
+        :rtype: list
```

Comment:
Now that I see this - should this (and or the get_unique_software) be an iterator? How / what is the context in which this is used?

Reply:
This is used in context of automating ``module load test``. 

Also for some reason the list is not sorted when testing the output

```
Unique Software: ['gompi', 'numactl', 'GCCcore', 'flex', 'ncurses', 'OpenMPI', 'Anaconda3', 'bzip2', 'foss', 'gettext', 'GCC', 'libiconv', 'libtool', 'libffi', 'GMP', 'Tcl', 'libpciaccess', 'Python', 'xz', 'gdbm', 'lmod', 'Automake', 'util-macros', 'Autotools', 'EasyBuild', 'libxml2', 'XZ', 'binutils', 'SQLite', 'tar', 'xorg-macros', 'OSU-Micro-Benchmarks', 'Autoconf', 'settarg', 'libreadline', 'OpenBLAS', 'ScaLAPACK', 'CUDA', 'CMake', 'FFTW', 'Clang', 'PyCharm', 'hwloc', 'Bison', 'm4', 'pkgconf', 'zlib', 'M4', 'libsigsegv', 'readline', 'help2man', 'diffutils']

```

so ``list(self.spider_content.keys()`` is not sorted.